Dismiss the First-Party Mode Setup Banner Notification When Toggle Is Enabled from Settings Screen

techanvil opened the pull request:
## Summary

<!-- Please reference the issue this PR addresses in the following list. -->
Addresses issue:

- #9698 

## Relevant technical choices

<!-- Please describe your changes. -->

## PR Author Checklist

- [x] My code is tested and passes existing unit tests.
- [x] My code has an appropriate set of unit tests which all pass.
- [x] My code is backward-compatible with WordPress 5.2 and PHP 7.4.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [x] My code has proper inline documentation.
- [ ] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Code Reviewer Checklist

- [x] Run the code.
- [x] Ensure the acceptance criteria are satisfied.
- [x] Reassess the implementation with the IB.
- [x] Ensure no unrelated changes are included.
- [x] Ensure CI checks pass.
- [x] Check Storybook where applicable.
- [x] Ensure there is a QA Brief.
- [x] Ensure there are no unexpected significant changes to file sizes.

## Merge Reviewer Checklist

- [x] Ensure the PR has the correct target branch.
- [x] Double-check that the PR is okay to be merged.
- [x] Ensure the corresponding issue has a ZenHub release assigned.
- [x] Add a changelog message to the issue.

techanvil commented on assets/js/modules/ads/datastore/settings.test.js:
```diff
@@ -109,6 +139,216 @@ describe( 'modules/ads settings', () => {
 			} );
 		} );
+
+		it( 'should handle an error when sending a POST request to the FPM settings endpoint', async () => {
```
I added a couple of additional tests to extend the coverage for the existing behaviour.

techanvil commented on assets/js/modules/ads/datastore/settings.test.js:
```diff
@@ -109,6 +139,216 @@ describe( 'modules/ads settings', () => {
+			expect( fetchMock ).not.toHaveFetched( fpmSettingsEndpoint );
+		} );
+
+		it( 'should dismiss the FPM setup CTA banner when the FPM `isEnabled` setting is changed to `true`', async () => {
```
Here's where the coverage for the newly introduced behaviour begins.

techanvil commented on assets/js/modules/analytics-4/datastore/settings.test.js:
```diff
@@ -521,27 +571,266 @@ describe( 'modules/analytics-4 settings', () => {
+				expect( console ).toHaveErrored();
+			} );
+
+			it( 'should not send a POST request to the FPM settings endpoint when the toggle state is changed', async () => {
```
As above, I added a couple of additional tests to extend the coverage for the existing behaviour.

techanvil commented on assets/js/modules/analytics-4/datastore/settings.test.js:
```diff
@@ -521,27 +571,266 @@ describe( 'modules/analytics-4 settings', () => {
+				expect( fetchMock ).not.toHaveFetched( fpmSettingsEndpoint );
+			} );
+
+			it( 'should dismiss the FPM setup CTA banner when the FPM `isEnabled` setting is changed to `true`', async () => {
```
Again, here's where the coverage for the newly introduced behaviour begins.

techanvil commented on assets/js/modules/analytics-4/datastore/settings.test.js:
```diff
@@ -513,6 +533,36 @@ describe( 'modules/analytics-4 settings', () => {
+				} );
+			} );
+
+			it( 'should handle an error when sending a POST request to the FPM settings endpoint', async () => {
```
As above, I added a couple of additional tests to extend the coverage for the existing behaviour.

techanvil commented on tests/js/utils.js:
```diff
@@ -463,15 +463,17 @@ export const provideKeyMetricsUserInputSettings = (
 	registry,
 	extraData,
-	overwrite = false
+	{ overwrite = false } = {}
```
It's preferable to avoid "unnamed" boolean parameters as they are not very clear when in use.

```js
			provideNotifications(
				registry,
				{
					[ FPM_SETUP_CTA_BANNER_NOTIFICATION ]:
						DEFAULT_NOTIFICATIONS[
							FPM_SETUP_CTA_BANNER_NOTIFICATION
						],
				},
				true, // What does `true` mean?
			);
```

benbowler commented:
FYI the PR wasn't linked to the issue in ZenHub, I just linked it.

benbowler reviewed (approved):
Thanks @techanvil, run, reviewed and moving to MR.

The use of constants for notification slugs makes sense, we should update other banners to use constants, ideally in the module constants for notifications related to a module or in this central file for others.

techanvil commented:
> Thanks @techanvil, run, reviewed and moving to MR.
> 
> The use of constants for notification slugs makes sense, we should update other banners to use constants, ideally in the module constants for notifications related to a module or in this central file for others.

Thanks @benbowler, yes, that would be good. I've created an issue to track this: https://github.com/google/site-kit-wp/issues/9835

hussain-t commented on assets/js/modules/analytics-4/datastore/settings.test.js:
```diff
@@ -521,27 +571,266 @@ describe( 'modules/analytics-4 settings', () => {
+				registry
+					.dispatch( CORE_SITE )
+					.receiveGetFirstPartyModeSettings( {
+						isEnabled: null,
```
Since we changed the `isEnabled` default value, let's update the mock as well.
```suggestion
						isEnabled: false,
```

hussain-t reviewed (changes requested):
Excellent work, @techanvil. I’ve updated the `isEnabled` default value to `false` in the test mocks to align with the recent change in the default settings value. However, the tests are failing due to the change. Could you please take a look at it?

techanvil commented:
> Excellent work, @techanvil. I’ve updated the `isEnabled` default value to `false` in the test mocks to align with the recent change in the default settings value. However, the tests are failing due to the change. Could you please take a look at it?

Thanks @hussain-t! I've fixed the tests. Back over to you :)

hussain-t reviewed (approved):
Excellent work, @techanvil! LGTM 👍 

Note: I've resolved the merged conflicts myself to save us from another round of iteration.
